scipy/optimize/_milp.py

```diff
@@ -0,0 +1,359 @@
+
+
+def _milp_iv(c, integrality, bounds, constraints, options):
+
+    # objective IV
+    c = np.atleast_1d(c).astype(np.double)
+    if c.ndim != 1:
+        message = "`c` must be a one-dimensional array."
+        raise ValueError(message)
+
+    # integrality IV
+    message = ("`integrality` must contain integers 0-3 and be broadcastable "
+               "to `c.shape`.")
+    if integrality is None:
+        integrality = 0
+    try:
+        integrality = np.broadcast_to(integrality, c.shape).astype(np.uint8)
+    except ValueError:
+        raise ValueError(message)
+    if integrality.min() < 0 or integrality.max() > 3:
+        raise ValueError(message)
+
+    # bounds IV
+    if bounds is None:
+        bounds = Bounds(0, np.inf)
+    elif not isinstance(bounds, Bounds):
+        message = "`bounds` must be an instance of `scipy.optimize.Bounds`."
+        try:
+            bounds = Bounds(*bounds)
+        except TypeError:
+            raise ValueError(message)
+
+    try:
+        lb = np.broadcast_to(bounds.lb, c.shape).astype(np.double)
+        ub = np.broadcast_to(bounds.ub, c.shape).astype(np.double)
+    except (ValueError, TypeError):
+        message = ("`bounds.lb` and `bounds.ub` must contain reals and "
+                   "be broadcastable to `c.shape`.")
+        raise ValueError(message)
+
+    # constraints IV
+    if not constraints:
+        constraints = [LinearConstraint(np.empty((0, c.size)),
+                                        np.empty((0,)), np.empty((0,)))]
+    A, b_l, b_u = _constraints_to_components(constraints)
+
+    if A.shape != (b_l.size, c.size):
+        message = "The shape of `A` must be (len(b_l), len(c))."
+        raise ValueError(message)
+    indptr, indices, data = A.indptr, A.indices, A.data.astype(np.double)
+
+    # options IV
+    options = options or {}
+    supported_options = {'disp', 'presolve', 'time_limit'}
+    unsupported_options = set(options).difference(supported_options)
+    if unsupported_options:
+        message = (f"Unrecognized options detected: {unsupported_options}. "
+                   "These will be passed to HiGHS verbatim.")
+        warnings.warn(message, RuntimeWarning, stacklevel=3)
+    options_iv = {'log_to_console': options.get("disp", False)}
+    options_iv.update(options)
+
+    return c, integrality, lb, ub, indptr, indices, data, b_l, b_u, options_iv
+
+
+def milp(c, *, integrality=None, bounds=None, constraints=None, options=None):
+    r"""
+    Mixed-integer linear programming
+
+    Solves problems of the following form:
+
+    .. math::
+
+        \min_x \ & c^T x \\
+        \mbox{such that} \ & b_l \leq A x \leq b_u,\\
+        & l \leq x \leq u, \\
+        & x_i \in \mathbb{Z}, i \in X_i
+
+    where :math:`x` is a vector of decision variables;
+    :math:`c`, :math:`b_l`, :math:`b_u`, :math:`l`, and :math:`u` are vectors;
+    :math:`A` is a matrix, and :math:`X_i` is the set of indices of
+    decision variables that must be integral.
```

Comment:
The term _integral_ might confuse people. It might be worth explaining it further in this function.

Reply:
```suggestion
    decision variables that must be integers.
```

Reply:
It would partially since we still have a param named `integrality`. I would just add a line explaining that integral means integer.

Reply:
```suggestion
    decision variables that must be integral. (Note that in this context, a
    variable that can only assume integer values is said to be "integral";
    it has an "integrality" constraint.)
```